### ISSUE TITLE
MPS add incorrectly

### ISSUE THREAD

**Issue by thipokKub** [pytorch/pytorch#77988]
### 🐛 Describe the bug

Using `add_` function to add two tensor together get the wrong result

**Noted** think this might be one of the reason why none of the optimizer can reduce any loss. Also I have tested separately for loss gradient calculation, but that seems to be working just fine (so far)

Reproduce
```python
import torch
import numpy as np

def seed_everything(seed: int):
    # Ref: https://gist.github.com/ihoromi4/b681a9088f348942b01711f251e5f964
    import random, os
    import numpy as np
    import torch
    
    random.seed(seed)
    os.environ['PYTHONHASHSEED'] = str(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)
    torch.cuda.manual_seed(seed)
    torch.backends.cudnn.deterministic = True
    torch.backends.cudnn.benchmark = True
    
seed_everything(1337)

coeff = 0.5
base_tensor = np.random.rand(100, 100).astype(np.float32)
grad_tensor = np.random.rand(100, 100).astype(np.float32)

device = "cpu"
cpu_tensor = torch.from_numpy(base_tensor).to(device)
cpu_tensor.requires_grad = True
cpu_tensor.grad = torch.from_numpy(grad_tensor).to(device)

with torch.no_grad():
    cpu_tensor.add_(-coeff*cpu_tensor.grad)

device = "mps"
mps_tensor = torch.from_numpy(base_tensor).to(device)
mps_tensor.requires_grad = True
mps_tensor.grad = torch.from_numpy(grad_tensor).to(device)

with torch.no_grad():
    mps_tensor.add_(-coeff*mps_tensor.grad)
    
print(cpu_tensor.detach().cpu().numpy() - mps_tensor.detach().cpu().numpy())
```

output
```
[[0.33641216 0.07313202 0.18328188 ... 0.11149368 0.34446636 0.13360822]
 [0.3013041  0.34535012 0.26777124 ... 0.2755718  0.15865743 0.12133723]
 [0.20747674 0.1145643  0.06512094 ... 0.46439692 0.45993388 0.49859798]
 ...
 [0.48521554 0.46384603 0.39157224 ... 0.1278103  0.06765462 0.4654624 ]
 [0.4041669  0.21006885 0.4867934  ... 0.16138805 0.05978572 0.2490782 ]
 [0.26637957 0.30114034 0.27918386 ... 0.0728505  0.03475738 0.08684569]]
```

expected all values to be `0` or very close to `0`

### Versions

PyTorch version: 1.12.0.dev20220520
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 12.3 (arm64)
GCC version: Could not collect
Clang version: 13.1.6 (clang-1316.0.21.2)
CMake version: version 3.22.2
Libc version: N/A

Python version: 3.9.12 | packaged by conda-forge | (main, Mar 24 2022, 23:25:14)  [Clang 12.0.1 ] (64-bit runtime)
Python platform: macOS-12.3-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.22.3
[pip3] pytorch-lightning==1.6.3
[pip3] pytorch-metric-learning==1.3.0
[pip3] torch==1.12.0.dev20220520
[pip3] torchaudio==0.11.0
[pip3] torchinfo==1.6.6
[pip3] torchmetrics==0.8.2
[pip3] torchvision==0.12.0
[conda] numpy                     1.21.6                   pypi_0    pypi
[conda] pytorch-lightning         1.6.3                    pypi_0    pypi
[conda] pytorch-metric-learning   1.3.0                    pypi_0    pypi
[conda] torch                     1.12.0.dev20220520          pypi_0    pypi
[conda] torchaudio                0.11.0                   pypi_0    pypi
[conda] torchinfo                 1.6.6                    pypi_0    pypi
[conda] torchmetrics              0.8.2                    pypi_0    pypi
[conda] torchvision               0.12.0                   pypi_0    pypi

**Comment by thipokKub** [comment]
Nevermind, the unexpected behaviour is from `cpu` not `mps`. I was using `from_numpy` which the document indicate it will create a tensor buffer that shared memory space (I didn't catch this at first). This shared memory space seems to only be valid for `cpu` version, and nor `mps` version

I tried switching `cpu`, and `mps` order around without using `copy()`, and got expected result, but the `base_tensor` value is modified after `cpu` computation. So in my conclusion, using `copy()` with `from_numpy` is needed to have consistent behaviour on both `cpu`, and `mps`

```python
import torch
import time
import numpy as np

def seed_everything(seed: int):
    # Ref: https://gist.github.com/ihoromi4/b681a9088f348942b01711f251e5f964
    import random, os
    import numpy as np
    import torch
    
    random.seed(seed)
    os.environ['PYTHONHASHSEED'] = str(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)
    torch.cuda.manual_seed(seed)
    torch.backends.cudnn.deterministic = True
    torch.backends.cudnn.benchmark = True
    
seed_everything(1337)

coeff = 0.5
base_tensor = np.random.rand(100, 100).astype(np.float32)
grad_tensor = np.random.rand(100, 100).astype(np.float32)

device = "cpu"
cpu_tensor = torch.from_numpy(base_tensor.copy()).to(device) # Change this line
cpu_tensor.requires_grad = True
cpu_tensor.grad = torch.from_numpy(grad_tensor.copy()).to(device) # Change this line

with torch.no_grad():
    cpu_tensor.add_(-coeff*cpu_tensor.grad)
    
device = "mps"
mps_tensor = torch.from_numpy(base_tensor.copy()).to(device) # Change this line
mps_tensor.requires_grad = True
mps_tensor.grad = torch.from_numpy(grad_tensor.copy()).to(device) # Change this line

with torch.no_grad():
    mps_tensor.add_(-coeff*mps_tensor.grad)
    
print(cpu_tensor.detach().cpu().numpy() - mps_tensor.detach().cpu().numpy())
```
